File: Week1_3/Merge_Sort.java
import java.time.Instant;
import java.time.Duration;
public class Merge_Sort extends Sort
{
    public Merge_Sort(int size)
    {
        super(size);
    }

    public void sort( )
    {
        int begin = 0;
        int end = N - 1;
        sort_kernel( elems, begin, end );
    }

    private void sort_kernel(int[] arr, int begin, int end )
    {
        if ( begin == end )
            return;
        else if ( begin + 1 == end )
        {
            if ( elems[begin] > elems[end] )
                swap(begin,end);
        }
        else
        {
            int mid = ( begin + end ) / 2;
            sort_kernel( arr, begin, mid );
            sort_kernel( arr, mid+1, end );
            merge(arr, begin, mid, end );
        }

    }

    private void merge( int[] arr, int begin, int mid, int end )
    {
        int[] aux = new int[end-begin+1];
        int i = begin;
        int j = mid+1;
        int k = 0;
        while ( i <= mid || j <= end )
        {
            if ( i > mid ) aux[k++] = arr[j++];
            else if ( j > end ) aux[k++] = arr[i++];
            else if ( arr[i] <= arr[j] ) aux[k++] = arr[i++];
            else aux[k++] = arr[j++];
        }
        k = 0;
        while ( k < (end-begin)+1 )
        {
            arr[begin+k] = aux[k];
            k++;
        }
    }

    public static void main(String[] args)
    {
        int nTest = 16;
        int N = 1;
        for ( int i = 1; i <= nTest; i++ )
        {
            N = N * 2;
            Sort ss = new Merge_Sort(N);
            ss.initialize();

            Instant start = Instant.now();
            ss.sort();
            Instant end = Instant.now();
            Duration dur = Duration.between(start,end);

            System.out.printf("N = %d, Runtime = %d ms, isSorted? = %b\n", N, dur.toMillis(), ss.isSorted());
        }
    }

}

File: Week1_3/Merge_Sort.cpp
#include "Sort.h"
#include <iostream>
#include <vector>
#include <fstream>
using namespace std;

class Merge_Sort : public Sort
{
    public:
        Merge_Sort(int size) : Sort(size) {};
        virtual void sort();
        void sort( vector<int> &arr, unsigned begin, unsigned end );
        void merge( vector<int> &arr, unsigned begin, unsigned mid, unsigned end);
};

void Merge_Sort::sort()
{
    unsigned begin = 0;
    unsigned end = _elems.size()-1;
    sort( _elems, begin, end );
}

void Merge_Sort::sort( vector<int> &arr, unsigned begin, unsigned end )
{
    if ( begin == end )
    {  return; }
    else if ( begin + 1 == end )
    {
        if ( arr[begin] > arr[end] )
            swap( arr[begin], arr[end] );
    }
    else
    {
        unsigned mid = ( begin + end ) / 2;
        sort( arr, begin, mid );
        sort( arr, mid+1, end );
        merge( arr, begin, mid, end);
    }
}

void Merge_Sort::merge( vector<int> &arr, unsigned begin, unsigned mid, unsigned end )
{
    unsigned i = begin;
    unsigned j = mid + 1;
    vector<int> aux;
    aux.reserve(end-begin+1);
    while ( i <= mid || j <= end )
    {
        if ( i > mid ) aux.push_back(arr[j++]);
        else if ( j > end ) aux.push_back(arr[i++]);
        else if ( arr[i] > arr[j] ) aux.push_back(arr[j++]);
        else aux.push_back(arr[i++]);
    }

    unsigned k = 0;
    while ( k < aux.size() )
    {
        arr[begin+k] = aux[k];
        k++;
    }
}


int main()
{
    unsigned Ntest = 20;
    unsigned N = 1;
    ofstream outFile("Merge_Sort_runtime.csv");
    outFile << "N,t" << endl;
    for ( unsigned itest = 0; itest < Ntest; itest++ )
    {
        N = N*2;
        Sort *ss = new Merge_Sort(N);
        ss->initialize();

        clock_t begin = clock();
        // ss->displayme();
        ss->sort();
        // ss->displayme();

        clock_t end = clock();

        cout << "N = " << ss->size() << ", RunTime: " <<  end-begin << ", isSorted?: " << ss->isSorted()  << endl;
        outFile << N << ", " << end-begin << endl;
    }
}
